src/ngraph/attribute_visitor.hpp

```diff
@@ -27,136 +28,125 @@ namespace ngraph
-        {
-            on_adapter(name, static_cast<ValueAccessor<void>&>(adapter));
-        }
+                                ValueAccessor<std::vector<float>>& adapter);
         virtual void on_adapter(const std::string& name,
-                                ValueAccessor<std::vector<std::string>>& adapter)
-        {
-            on_adapter(name, static_cast<ValueAccessor<void>&>(adapter));
-        }
-        // Use an adapter for non-primitive types
-        template <typename T>
-        // typename std::enable_if<std::is_class<T>::value, void>::type
-        void on_attribute(const std::string& name, T& value)
-        {
-            AttributeAdapter<T> adapter(value);
-            on_adapter(name, adapter);
-        }
-        void on_attribute(const std::string& name, op::AutoBroadcastSpec& value)
-        {
-            AttributeAdapter<op::AutoBroadcastType> adapter(value.m_type);
-            on_adapter(name, adapter);
-        }
-        void on_attribute(const std::string& name, op::BroadcastModeSpec& value)
-        {
-            AttributeAdapter<op::BroadcastType> adapter(value.m_type);
+                                ValueAccessor<std::vector<double>>& adapter);
+        virtual void on_adapter(const std::string& name,
+                                ValueAccessor<std::vector<std::string>>& adapter);
+        /// \brief Hook for inline struct visit
+        virtual void on_adapter(const std::string& name, StructAdapter& adapter);
+
+        /// Legacy method
+        virtual void on_attribute(const std::string& name, std::string& value)
+            NGRAPH_DEPRECATED("Override on_adapter instead");
+        /// Legacy method
+        virtual void on_attribute(const std::string& name, bool& value)
+            NGRAPH_DEPRECATED("Override on_adapter instead");
+        /// The generic visitor. There must be a definition of AttributeAdapter<T> that can convert
+        /// to a ValueAccessor<U> for one of the on_adpater methods.
+        template <typename AT>
+        void on_attribute(const std::string& name, AT& value)
+        {
+            AttributeAdapter<AT> adapter(value);
             on_adapter(name, adapter);
         }
+        /// \returns The nested context of visits
+        const std::vector<Context>& get_context() const { return m_context; }
+        /// \returns context prepended to names
+        virtual std::string get_name_with_context(const std::string& name);
+        /// \brief Start visiting a nested structure
+        virtual void start_structure(const std::string& name);
+        /// \brief Finish visiting a nested structure
+        virtual void finish_structure();
+        /// \brief Get the id for a pointer
+        virtual int64_t get_id(void*);
+        /// \brief Retrieve a pointer from the id table
+        virtual void* get_ptr(int64_t id);
+        /// \brief Add a pointer to the id table
+        virtual int64_t insert_ptr(void*);
```

Comment:
What are those methods for? Some usage scenario example would be helpful.

Reply:
Since nodes are deserialized a node at a time from formats that have their own way to interconnect nodes, this (slightly different now) is a way for the serialization format to associate its ids with previously processed nodes.